docs/presentations/pycon2024/maintainers.qmd

```diff
@@ -0,0 +1,322 @@
+```
+
+## 16+ other things
+
+![](./machine.gif){fig-align="center" width="100%" height="100%"}
+
+:::
+
+## how it works
+
+Ibis compiles down to SQL or dataframe code:
+
+```{python}
+#| echo: false
+
+import os
+import sys
+sys.path.append(os.path.abspath(".."))
+
+from backends_sankey import fig
+fig.show()
+```
+
+# what's in Ibis?
+
+## by the numbers {.smaller}
+
+:::: {.columns}
+::: {.column width="50%"}
+### backends
+- **17** SQL
+- **3** non-SQL
+- **2** cloud
+:::
+
+::: {.column width="50%"}
+### engines + APIs
+- **9** distributed SQL
+- **3** dataframe
+- oldest: **~45** years 👀
+- newest: **~2** years
+:::
+::::
+
+### other facts
+
+- latency is wild
+- deployment models vary
+
+::: {.fragment}
+::: {.r-fit-text}
+_… **feature development**_❓
+:::
+:::
+
+## bit of a pickle
+
+![](./picklerick.png)
+
+# how
+
+## high level
+
+### goal: fast iteration
+
+- fast env setup (dep mgmt)
+- (less) fast tests (testing)
+- high **job** concurrency (ci/provider)
+- **easy to run**: dev speed ([`just`](https://github.com/casey/just))
+
+::: {.fragment}
+::: {.r-fit-text}
+_CI must complete "quickly"_
+:::
+:::
+
+## tools: overview
+
+- **deps**: poetry
+- **ci**: GitHub Actions
+- **beasts**: docker
+- **house pets**: docker
+- cool kids don't get special tx (duckdb, polars)
+- task runner (e.g.: `just up postgres`)
+
+## tools: poetry
+
+::: {.callout-warning}
+## Opinions follow
+:::
+
+- **env setup needs to be _fast_**: avoid SAT
+- poetry is one way; there are others
+- get yourself a lockfile
+
+::: {.fragment}
+::: {.r-fit-text}
+… _are you doing that **now**_❔❓
+:::
+:::
+
+## tools: docker
+
+- Sure, docker
+- But, do you to use it locally?
+- Use health checks; "dumb" ones are fine
+- Focus: make it easy for devs to use
+
+## tools: GitHub actions
+
+- Pay for the Teams plan ([more concurrency](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits))
+- Automate dep updates via Renovate/Dependabot
+
+::: {.columns}
+::: {.column width="50%"}
+### GHA concurrency limits
+
+![](./gha.png)
+:::
+
+::: {.column width="50%"}
+### ibis CI cost
+
+![](./bill.png)
+:::
+:::
+
+## pytest {.smaller}
+
+### Ibis problems
+
+- Backends don't implement the same stuff
+- Need to know when backend passes
+- Need to specify exception type it raises
+- Answer questions like: "will it _ever_ blend?"
+
+::: {.fragment}
+### markers + hooks
+
+```python
+@pytest.mark.never("duckdb")  # never gonna happen
+@pytest.mark.notyet("impala")  # might happen
+@pytest.mark.notimpl("snowflake")  # ibis devs: do some work
+def test_soundex():
+    ...
+
+def pytest_ignore_collect(...):
+    # pytest -m duckdb: don't collect things that aren't marked duckdb
+    ...
+```
+:::
+
+## pytest: plugins
+
+**`pytest-`**
+
+- `xdist`: try to make this work if you can
+- `randomly`: exposes your bogus and stateful assumptions
+- `repeat`: great when randomly has exposed your shenanigans
```

Comment:
exposed -> exposes? I also don't really understand what this means but assume you'll explain it

Reply:
i will definitely explain this in person

Reply:
changed to

> exposes